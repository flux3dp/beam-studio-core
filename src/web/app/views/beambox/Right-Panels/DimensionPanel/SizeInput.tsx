import React, { memo, useCallback, useMemo } from 'react';

import ObjectPanelItem from 'app/views/beambox/Right-Panels/ObjectPanelItem';
import UnitInput from 'app/widgets/Unit-Input-v2';
import storage from 'implementations/storage';
import { useIsMobile } from 'helpers/system-helper';

import styles from './DimensionPanel.module.scss';

interface Props {
  type: 'w' | 'h' | 'rx' | 'ry';
  value: number;
  onChange: (type: string, value: number) => void;
  onBlur?: () => void;
  onKeyUp?: (e: KeyboardEvent) => void;
}

const SizeInput = ({ type, value, onChange, onBlur, onKeyUp }: Props): JSX.Element => {
  const isMobile = useIsMobile();
  const unit = useMemo(() => storage.get('default-units') === 'inches' ? 'in' : 'mm', []);

  const label = useMemo<string | JSX.Element>(() => {
    if (type === 'w') return 'W';
    if (type === 'h') return 'H';
    if (type === 'rx') return 'W';
    if (type === 'ry') return 'H';
    return null;
  }, [type]);
  const handleChange = useCallback((val: number) => {
    const changeKey = {
      w: 'width',
      h: 'height',
      rx: 'rx',
      ry: 'ry',
    }[type];
    const newVal = type === 'rx' || type === 'ry' ? val / 2 : val;
    onChange(changeKey, newVal);
  }, [onChange, type]);

  if (isMobile) {
    return (
      <ObjectPanelItem.Number
        id={`${type}_size`}
        value={value}
        updateValue={handleChange}
        label={label}
      />
    );
  }
  return (
    <div className={styles.dimension}>
      <div className={styles.label}>{label}</div>
      <UnitInput
        id={`${type}_size`}
        unit={unit}
        onBlur={onBlur}
        onKeyUp={onKeyUp}
        defaultValue={value}
        getValue={handleChange}
        min={0}
      />
    </div>
  );
};

export default memo(SizeInput);
