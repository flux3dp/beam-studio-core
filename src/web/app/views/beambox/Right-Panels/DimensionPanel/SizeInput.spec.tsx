import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import SizeInput from './SizeInput';

const mockGet = jest.fn();
jest.mock('implementations/storage', () => ({
  get: (...args) => mockGet(...args),
}));

const mockUseIsMobile = jest.fn();
jest.mock('helpers/system-helper', () => ({
  useIsMobile: () => mockUseIsMobile(),
}));

jest.mock(
  'app/widgets/Unit-Input-v2',
  () =>
    ({ id, unit, defaultValue, getValue, onBlur, onKeyUp }: any) =>
      (
        <div>
          <div>{id}</div>
          <div>{unit}</div>
          <div>{defaultValue}</div>
          <div>{getValue}</div>
          <input
            type="number"
            defaultValue={defaultValue}
            onChange={(e) => getValue(Number(e.target.value))}
            onBlur={onBlur}
            onKeyUp={onKeyUp}
          />
        </div>
      )
);

jest.mock('app/views/beambox/Right-Panels/ObjectPanelItem', () => ({
  Number: ({ id, value, updateValue, label }: any) => (
    <div id={id}>
      {label}
      <div>{value}</div>
      <button type="button" onClick={() => updateValue(value + 1)} />
    </div>
  ),
}));

const mockOnChange = jest.fn();
const mockOnBlur = jest.fn();
const mockOnKeyUp = jest.fn();

describe('test SizeInput', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly on desktop', () => {
    mockUseIsMobile.mockReturnValue(false);
    const { container } = render(<SizeInput type="w" value={0} onChange={mockOnChange} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly on mobile', () => {
    mockUseIsMobile.mockReturnValue(true);
    const { container } = render(<SizeInput type="w" value={0} onChange={mockOnChange} />);
    expect(container).toMatchSnapshot();
  });

  test('onChange on desktop', () => {
    mockUseIsMobile.mockReturnValue(false);
    const { container } = render(<SizeInput type="w" value={0} onChange={mockOnChange} />);
    const input = container.querySelector('input');
    fireEvent.change(input, { target: { value: 1 } });
    expect(mockOnChange).toBeCalledTimes(1);
    expect(mockOnChange).toHaveBeenLastCalledWith('width', 1);
  });

  test('onChange on mobile', () => {
    mockUseIsMobile.mockReturnValue(true);
    const { container } = render(<SizeInput type="w" value={0} onChange={mockOnChange} />);
    const button = container.querySelector('button');
    fireEvent.click(button);
    expect(mockOnChange).toBeCalledTimes(1);
    expect(mockOnChange).toHaveBeenLastCalledWith('width', 1);
  });

  test('onBlur', () => {
    mockUseIsMobile.mockReturnValue(false);
    const { container } = render(
      <SizeInput type="w" value={0} onChange={mockOnChange} onBlur={mockOnBlur} />
    );
    const input = container.querySelector('input');
    fireEvent.blur(input);
    expect(mockOnBlur).toBeCalledTimes(1);
  });

  test('onKeyUp', () => {
    mockUseIsMobile.mockReturnValue(false);
    const { container } = render(
      <SizeInput type="w" value={0} onChange={mockOnChange} onKeyUp={mockOnKeyUp} />
    );
    const input = container.querySelector('input');
    fireEvent.keyUp(input);
    expect(mockOnKeyUp).toBeCalledTimes(1);
  });
});
