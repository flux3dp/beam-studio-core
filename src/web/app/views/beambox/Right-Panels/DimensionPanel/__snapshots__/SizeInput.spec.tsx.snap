// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SizeInput should render correctly on desktop 1`] = `
<div>
  <div
    class="dimension"
  >
    <div
      class="label"
    >
      W
    </div>
    <div>
      <div>
        w_size
      </div>
      <div>
        mm
      </div>
      <div>
        0
      </div>
      <div />
      <input
        type="number"
        value="0"
      />
    </div>
  </div>
</div>
`;

exports[`test SizeInput should render correctly on mobile 1`] = `
<div>
  <div
    id="w_size"
  >
    W
    <div>
      0
    </div>
    <button
      type="button"
    />
  </div>
</div>
`;
